### PR TITLE
Issue #1971 Add-on to change the value of 'max-pods' in node config

### LINKED ISSUE #1971: Change limit pod/cpu
Statement: 

"I experiment from time to time such problem on my machine using "oc cluster up"  (openshift origin 3.7.x)

`0/1 nodes are available: 1 Insufficient pods.`

If I look to the to stats generated by the command "oc get node localhost", then I can see that the limit of the pods created has been reached.

As the convention followed by openshift is to limit the number of the pods "available" to 10pods/cpu, I'm wondering how we could change this value when a cluster is started ?"

According to the openshift team, such limit can be managed using this parameter

"this limit can be changed in the node configuration, see: https://docs.openshift.com/container-platform/3.7/admin_guide/manage_nodes.html#admin-guide-max-pods-per-node
Looking at https://github.com/openshift/origin/blob/master/docs/cluster_up_down.md#configuration it seems that the configuration file should be available under /var/lib/origin/openshift.local.config with oc cluster up. You can use --use-existing-config argument when restarting your cluster to avoid regenerating the configuration."

### PR DESCRIPTION
**Pull request by anjannath**
Fixes #1971

### CODE REVIEW

File: addons/max-pod/max-pod.addon [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+# Required-Vars: MAX_PODS  
+# OpenShift-Version: >3.6.0
+# Var-Defaults: MAX_PODS=20
+
+
+ssh docker exec -t origin /usr/bin/openshift ex config patch /var/lib/origin/openshift.local.config/node-localhost/node-config.yaml '--patch={"kubeletArguments": { "max-pods": ["#{MAX_PODS}"] } }' > tmpconfig
+
+ssh cat tmpconfig | sed "s/'/\'/" | sed "s/.$//" | sudo tee /var/lib/minishift/openshift.local.config/node-localhost/node-config.yaml > /dev/null
+ssh rm tmpconfig
+
```
Comment (LalatenduMohanty) [inline addons/max-pod/max-pod.addon]: @anjannath you can directly use `docker exec` command like https://github.com/minishift/minishift-addons/blob/master/add-ons/cors/cors.addon.

File: addons/max-pod/max-pod.addon [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+docker stop origin
+docker start origin
```
Comment (LalatenduMohanty) [inline addons/max-pod/max-pod.addon]: docker restart origin

**Comment by praveenkumar** [comment]
And why this is part of this repo instead `minishift-addon` one?

File: addons/max-pod/max-pod.addon (hunk 1 above)
Comment (LalatenduMohanty) [reply addons/max-pod/max-pod.addon]: Also I think you can remove the intermediate file and directly apply the sed.

**Comment by LalatenduMohanty** [comment]
> And why this is part of this repo instead minishift-addon one?

IMO the issue is part of core Minishift and the solution should be part of core Minishift. Not sure what others think @minishift/minishift-dev .

**Comment by anjannath** [comment]
The reason for it not in the `add-on` repo because, it is not something like an application, its more similar to the add-ons in this repo like the `admin-user` or `anyuid` where we are changing openshift configs. 

That being said, i have a bit of hesitation to make it an addon and want people to use the commad line `minishift openshift config --patch` to patch the node config file. The command line takes care of backing up the previous config file and restores it if anything goes wrong. In the addon we are not doing that :(. And we might end up with a corrupt config file, whic would stop openshift from starting.

File: addons/max-pod/max-pod.addon (hunk 1 above)
Comment (anjannath) [reply addons/max-pod/max-pod.addon]: Thanks @LalatenduMohanty i will try with that

File: addons/max-pod/max-pod.addon [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Name: max-pod
+# Description: Add-on to change the value of max-pods
```
Comment (budhrg) [inline addons/max-pod/max-pod.addon]: `Add-on to update the maximum pod value`

Comment (anjannath) [reply addons/max-pod/max-pod.addon]: @budhrg Thanks.. updating the description, WDYT about the name?

Comment (budhrg) [reply addons/max-pod/max-pod.addon]: Name is fine. :+1:

**Comment by LalatenduMohanty** [comment]
@anjannath  The add-on is not part of defaults as `minishift addons install --defaults` does not installs it.

**Comment by anjannath** [comment]
@LalatenduMohanty Yeah, sorry i forgot to add it to defaultAssests, can you please try again, It should install now.

File: addons/max-pod/max-pod.addon (hunk 1 above)
Comment (anjannath) [reply addons/max-pod/max-pod.addon]: @LalatenduMohanty we need to use `ssh` as the patching process interects both with the container and the host vm.

**Comment by LalatenduMohanty** [comment]
@anjannath there seems to be issue with the default install and some kind of bug in general

```
$ rm -rf ~/.minishift/addons

$ minishift addons list
Error installing default add-ons : Asset max-pod not found

$ minishift addons list
- admin-user     : disabled   P(0)
- anyuid         : disabled   P(0)
- registry-route : disabled   P(0)
- xpaas          : disabled   P(0)
```
`Error installing default add-ons : Asset max-pod not found` is bug and we need to fix it.

```
$ minishift addon install --defaults
Default add-ons 'anyuid, admin-user, xpaas, registry-route, max-pod' installed

$ minishift addons list
- admin-user     : disabled   P(0)
- anyuid         : disabled   P(0)
- registry-route : disabled   P(0)
- xpaas          : disabled   P(0)

```
Still no sign of `max-pod`.

**Comment by anjannath** [comment]
@LalatenduMohanty It should have worked, also the ci didn't fail (made changes to the feature file to have 'max-pod' as default addon).
Did you do `make clean` before building this branch? you need to, as `go-bindata` needs to pack the addon.

**Comment by budhrg** [comment]
> Did you do make clean before building this branch? you need to, as go-bindata needs to pack the addon.

I was also getting similar error as @LalatenduMohanty is getting. 
Doing `make clean` and `make`, I am not getting any error.

File: addons/max-pod/max-pod.addon [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Name: max-pod
+# Description: Add-on to update the maximum pods value
```
Comment (budhrg) [inline addons/max-pod/max-pod.addon]: Could you add value as well here in the description so that it is available in the verbose view.

**Comment by budhrg** [comment]
Functionally working. Able to see the node's configuration to have max pod
```
$ minishift openshift config view --target node
allowDisabledDocker: false
apiVersion: v1
[...]
imageConfig:
  format: openshift/origin-${component}:v3.7.1
  latest: false
iptablesSyncPeriod: 30s
kind: NodeConfig
kubeletArguments:
  cgroup-driver:
  - cgroupfs
  max-pods:
  - "20"
masterClientConnectionOverrides:
  acceptContentTypes: application/vnd.kubernetes.protobuf,application/json
  burst: 40
  contentType: application/vnd.kubernetes.protobuf
  qps: 20
[...]
```

File: addons/max-pod/max-pod.addon (hunk 4 above)
Comment (anjannath) [reply addons/max-pod/max-pod.addon]: I could, but the user is supposed to pass that value, the add-on is to allow the user to change the 'max-pods' value, not set it to a fixed 20. maybe we should just remove the `Default-Vars` ? WDYT?

Comment (budhrg) [reply addons/max-pod/max-pod.addon]: Then, updated description as 
```
Add-on to update the maximum pods value (default to 20)
```

Comment (budhrg) [reply addons/max-pod/max-pod.addon]: Also, need documentation update.

File: docs/source/using/addons.adoc [hunk 5]
```diff
@@ -255,6 +255,8 @@ By default, this is not allowed in OpenShift.
 | Creates a user named 'admin' and assigns the *cluster-admin* role to it.  
 | registry-route
 | Creates an edge terminated route for the OpenShift registry.
+| max-pod
```
Comment (anjannath) [inline docs/source/using/addons.adoc]: Hi @robin-owen can you please take a look, i am not very satisfied with the description :-P

Comment (robin-owen) [reply docs/source/using/addons.adoc]: What's there is true, but doesn't help the reader understand what the add-on does. Something along the lines of "adjusts the maximum number of pods allocated for each CPU core" could work. :)

Comment (anjannath) [reply docs/source/using/addons.adoc]: Aha.. now there's a shortcoming in this addon, to set the number of pods/CPU core, 'pods-per-core' value has to be modified. both the properties have an effect on the number of pods a node can run.

ref: https://docs.openshift.com/container-platform/3.5/admin_guide/manage_nodes.html#admin-guide-max-pods-per-node

**Comment by LalatenduMohanty** [comment]
@anjannath  I can not move past 18 pods even of I used 6GB for the VM and max pods was 30. 
 
```
$ minishift start --memory 6GB

$ minishift addon apply  --addon-env MAX_PODS=30 max-pod

$ minishift openshift config view --target node
allowDisabledDocker: false
apiVersion: v1
authConfig:
  authenticationCacheSize: 1000
  authenticationCacheTTL: 5m
  authorizationCacheSize: 1000
  authorizationCacheTTL: 5m
dnsBindAddress: ""
dnsDomain: cluster.local
dnsIP: 172.30.0.1
dnsNameservers: null
dnsRecursiveResolvConf: ""
dockerConfig:
  dockerShimRootDirectory: /var/lib/dockershim
  dockerShimSocket: /var/run/dockershim.sock
  execHandlerName: native
enableUnidling: true
imageConfig:
  format: openshift/origin-${component}:v3.7.1
  latest: false
iptablesSyncPeriod: 30s
kind: NodeConfig
kubeletArguments:
  cgroup-driver:
  - systemd
  max-pods:
  - "30"
```

image


image

File: addons/max-pod/max-pod.addon [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Name: max-pod
+# Description: Add-on to update the maximum pods value (defaults to 20)
```
Comment (budhrg) [inline addons/max-pod/max-pod.addon]: Looks good now

**Comment by LalatenduMohanty** [comment]
@anjannath I would suggest you to close this PR and once you are sure about the issue please re-open the PR with required changes.

**Pull request closed without merging**